NOISSUE - Use Nats JetStream As Internal Message Bus

rodneyosodo opened the pull request:
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

### What does this do?
Moves the internal message bus implementation to use Nats JetStream.

### Which issue(s) does this PR fix/relate to?
No issue

### List any changes that modify/break current functionality
Update the `NewPublisher` and `NewPubSub` function parameters.

### Have you included tests for your changes?
I have added integration tests to ensure the new Nats JetStream implementation works correctly.

### Did you document any new/modified functionality?
Not yet

### Notes
This change improves the performance and scalability of our internal message bus by switching to Nats JetStream. The new implementation maintains the same external interface, so there should be minimal disruption.

Nats provides 2 APIs to interact with Jestream. `jetstream` package is preferred compared to the `nats` package because,
> `jetstream` package is a new client API to interact with NATS JetStream, aiming to replace the JetStream client implementation from `nats` package.

To be merged after https://github.com/mainflux/mainflux/pull/1885

rodneyosodo committed:
Replace Nats with Nats Jestream For PubSub

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Add Stream Description

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Subscribe using wildcard

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Add consumers description

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Remove unused queue variable

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

SammyOina commented on docker/.env:
```diff
@@ -377,7 +379,7 @@ MF_POSTGRES_HOST=mainflux-postgres
 MF_POSTGRES_USER=mainflux
 MF_POSTGRES_PASS=mainflux
-MF_POSTGRES_NAME=mainflux
+MF_POSTGRES_NAME=messages
```
is this change required?

rodneyosodo replied:
Yes.

rodneyosodo committed:
Add extra configs to stream

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

arvindh123 reviewed (commented):
LGTM

rodneyosodo committed:
Use inline error handling

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Fix connection leak in NATS publisher

The publisher struct in pkg/messaging/nats/publisher.go was modified to include a new `conn` field of type `*broker.Conn`. This change was made to fix a connection leak issue in the NATS publisher.

The `NewPublisher` function was updated to assign the `conn` parameter to the new `conn` field in the publisher struct.

Additionally, the `Close` method in the publisher struct was modified to close the `conn` connection.

This commit fixes the connection leak issue in the NATS publisher and ensures that connections are properly closed.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

dborovcanin commented on pkg/messaging/nats/pubsub.go:
```diff
@@ -143,27 +169,35 @@ func (ps *pubsub) Unsubscribe(ctx context.Context, id, topic string) error {
+
+	current.Stop()
+
+	if err := ps.stream.DeleteConsumer(ctx, id); err != nil {
```
Unlike plain NATS, JS client is able to hold consumer (subscription) data. Do we need to hold our own list of subscriptions in that case?

rodneyosodo replied:
I have made the consumer durable meaning consumers can have subscriptions bound to the topic and resume until the consumer is explicitly deleted. We keep an internal list of subscriptions to track the `ConsumeContext` across consumers so that we can stop and delete the consumer upon unsubscribing

rodneyosodo committed:
feat(messaging): Add support for durable consumers

This commit adds support for durable consumers to NATS JS in the messaging package.

To support this functionality, the `strings` package has been imported in the `pubsub.go` file to check the topic.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
feat : remove internal logic to keep subscribers

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on pkg/messaging/nats/pubsub.go:
```diff
@@ -126,44 +116,42 @@ func (ps *pubsub) Unsubscribe(ctx context.Context, id, topic string) error {
+	topic = strings.ReplaceAll(topic, "\\", "_")
+
+	return fmt.Sprintf("%s-%s", topic, id)
+}
```
Simplify:
```go
func formatConsumerName(topic, id string) string {
	// A durable name cannot contain whitespace, ., *, >, path separators (forward or backwards slash), and non-printable characters.
	chars := []string{
		" ", "_",
		".", "_",
		"*", "_",
		">", "_",
		"/", "_",
		"\\", "_",
	}
	rplcr := strings.NewReplacer(chars...)
	topic = rplcr.Replace(topic)
	return fmt.Sprintf("%s-%s", topic, id)
}
```

rodneyosodo committed:
feat(messaging): add function to format consumer name

This commit adds a new function to the `pubsub` package in the `messaging` module. The function is called `formatConsumerName` and is used to generate a consumer name for NATS messaging. The function takes two parameters, `topic` and `id`, and returns a formatted consumer name. The consumer name is generated by concatenating the `topic` and `id` parameters, with some restrictions on the characters that can be used. This function will be useful for creating durable subscriptions in NATS messaging.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

drasko reviewed (approved):
LGTM
